### PR TITLE
Add all negative cases for None and WildFly runtimes

### PR DESCRIPTION
**Pull request by Tasnim1147**

### CODE REVIEW

**Review by m-reza-rahman — DISMISSED** [review]
I am having trouble understanding this PR. Why are these changes needed and how are they related to negative cases? Is this still WIP and marked for review by mistake?

**Comment by Tasnim1147** [comment]
Here is a summary of the changes I have made. I have added negative test cases for None and WildFly runtimes. There are two types of negative cases.

**First type:** The build should be successful, but the Dockerfile should not be generated. I have tested for the absence of the Dockerfile. For example:
```
 name: Run Archetype for EE 8, SE 8, with Docker
        run: |
          mvn archetype:generate -DinteractiveMode=false -DaskForDefaultPropertyValues=false -DarchetypeGroupId=org.eclipse.starter -DarchetypeArtifactId=jakarta-starter -DarchetypeVersion=2.3.0-SNAPSHOT -DjakartaVersion=8 -Dprofile=full -DjavaVersion=8 -Ddocker=yes -DoutputDirectory=app -Dgoals="clean package"
          if [ -f app/jakartaee-hello-world/Dockerfile ]; then
            echo "Dockerfile found, test failed."
            exit 1
          fi
          rm -rf app
```
**Second type:** The build should fail. I have tested that the Maven build is not successful. For example:
```
- name: Run Archetype for EE 8 Core Profile, SE 8
        id: ee_8_core_se_8_none
        continue-on-error: true
        run: |
          mvn archetype:generate -DinteractiveMode=false -DaskForDefaultPropertyValues=false -DarchetypeGroupId=org.eclipse.starter -DarchetypeArtifactId=jakarta-starter -DarchetypeVersion=2.3.0-SNAPSHOT -DjakartaVersion=8 -Dprofile=core -DjavaVersion=8 -DoutputDirectory=app -Dgoals="clean package"
      - name: Check Maven Build Status
        run: |
          if [[ "${{ steps['ee_8_core_se_8_none'].outcome }}" != "failure" ]]; then
            echo "Maven build for EE 8 Core Profile, SE 8, without Docker succeeded, test failed."
            exit 1
          fi
```

**Comment by m-reza-rahman** [comment]
Thanks, I got it. Likely just a GitHub glitch. I'll try to review over the weekend.

File: .github/workflows/nightly.yml [hunk 1]
```diff
@@ -31,31 +31,241 @@ jobs:
...
+          mvn archetype:generate -DinteractiveMode=false -DaskForDefaultPropertyValues=false -DarchetypeGroupId=org.eclipse.starter -DarchetypeArtifactId=jakarta-starter -DarchetypeVersion=2.3.0-SNAPSHOT -DjakartaVersion=8 -Dprofile=core -DjavaVersion=8 -DoutputDirectory=app -Dgoals="clean package"
+
+      - name: Check Maven Build Status
+        run: |
+          if [[ "${{ steps['ee_8_core_se_8_none'].outcome }}" != "failure" ]]; then
+            echo "Maven build for EE 8 Core Profile, SE 8, without Docker succeeded, test failed."
+            exit 1
+          fi
+
+      - name: Run Archetype for EE 8 Core Profile, SE 8, with Docker
```
Comment (m-reza-rahman) [inline .github/workflows/nightly.yml:70]: Since the previous test checks for failure, there is no need for this.

File: .github/workflows/nightly.yml [hunk 2]
```diff
@@ -31,31 +31,241 @@ jobs:
...
+          mvn archetype:generate -DinteractiveMode=false -DaskForDefaultPropertyValues=false -DarchetypeGroupId=org.eclipse.starter -DarchetypeArtifactId=jakarta-starter -DarchetypeVersion=2.3.0-SNAPSHOT -DjakartaVersion=9 -Dprofile=core -DjavaVersion=8 -DoutputDirectory=app -Dgoals="clean package"
+
+      - name: Check Maven Build Status
+        run: |
+          if [[ "${{ steps['ee_9_core_se_8_none'].outcome }}" != "failure" ]]; then
+            echo "Maven build for EE 9 Core Profile, SE 8, without Docker succeeded, test failed."
+            exit 1
+          fi
+
+      - name: Run Archetype for EE 9 Core Profile, SE 8, with Docker
```
Comment (m-reza-rahman) [inline .github/workflows/nightly.yml:124]: Since the previous test checks for failure, there is no need for this.

File: .github/workflows/nightly.yml [hunk 3]
```diff
@@ -31,31 +31,241 @@ jobs:
           mvn archetype:generate -DinteractiveMode=false -DaskForDefaultPropertyValues=false -DarchetypeGroupId=org.eclipse.starter -DarchetypeArtifactId=jakarta-starter -DarchetypeVersion=2.3.0-SNAPSHOT -DjakartaVersion=8 -Dprofile=full -DjavaVersion=8 -DoutputDirectory=app -Dgoals="clean package"
           rm -rf app
 
+      - name: Run Archetype for EE 8, SE 8, with Docker
```
Comment (m-reza-rahman) [inline .github/workflows/nightly.yml:34]: Since we know the Dockerfile is really not generated based on runtime and not the other parameters, I think we can reduce the number of tests a little. I think lets do the Dockerfile check only once for the very first one for a runtime - so probably just keep this one and not the rest of it's kind.

File: .github/workflows/nightly.yml [hunk 4]
```diff
@@ -31,31 +31,241 @@ jobs:
...
+      - name: Run Archetype for EE 8 Web Profile, SE 8, with Docker
+        run: |
+          mvn archetype:generate -DinteractiveMode=false -DaskForDefaultPropertyValues=false -DarchetypeGroupId=org.eclipse.starter -DarchetypeArtifactId=jakarta-starter -DarchetypeVersion=2.3.0-SNAPSHOT -DjakartaVersion=8 -Dprofile=web -DjavaVersion=8 -Ddocker=yes -DoutputDirectory=app -Dgoals="clean package"
+          if [ -f app/jakartaee-hello-world/Dockerfile ]; then
+            echo "Dockerfile found, test failed."
+            exit 1
+          fi
+          rm -rf app
+
+      - name: Run Archetype for EE 8 Core Profile, SE 8
```
Comment (m-reza-rahman) [inline .github/workflows/nightly.yml:57]: Since we know the Core Profile selection will fail based on the Jakarta EE version and not the other parameters, I think we can reduce the number of tests a little. I think lets do this particular check only once for each Jakarta EE version - so probably just keep this one and not the rest of it's kind for a Jakarta EE version.

File: .github/workflows/nightly.yml [hunk 5]
```diff
@@ -31,31 +31,241 @@ jobs:
...
+          mvn archetype:generate -DinteractiveMode=false -DaskForDefaultPropertyValues=false -DarchetypeGroupId=org.eclipse.starter -DarchetypeArtifactId=jakarta-starter -DarchetypeVersion=2.3.0-SNAPSHOT -DjakartaVersion=9.1 -Dprofile=core -DjavaVersion=8 -Ddocker=yes -DoutputDirectory=app -Dgoals="clean package"
+
+      - name: Check Maven Build Status
+        run: |
+          if [[ "${{ steps['ee_9_1_core_se_8_none_docker'].outcome }}" != "failure" ]]; then
+            echo "Maven build for EE 9.1 Core Profile, SE 8, with Docker succeeded, test failed."
+            exit 1
+          fi
+
+      - name: Run Archetype for EE 10, SE 8
```
Comment (m-reza-rahman) [inline .github/workflows/nightly.yml:191]: Since we know the SE 8 selection will fail based on the Jakarta EE version (10) and not the other parameters, I think we can reduce the number of tests a little. I think lets do this particular check only once for EE 10 version - so probably just keep this one and not the rest of it's kind for EE 10.

File: .github/workflows/nightly.yml [hunk 6]
```diff
@@ -146,6 +356,10 @@ jobs:
       - name: Run Archetype for EE 8 Web Profile, SE 8, WildFly
         run: |
           mvn archetype:generate -DinteractiveMode=false -DaskForDefaultPropertyValues=false -DarchetypeGroupId=org.eclipse.starter -DarchetypeArtifactId=jakarta-starter -DarchetypeVersion=2.3.0-SNAPSHOT -DjakartaVersion=8 -Dprofile=web -DjavaVersion=8 -Druntime=wildfly -DoutputDirectory=app/wildfly -Dgoals="clean package"
+          if [ -f app/wildfly/jakartaee-hello-world/Dockerfile ]; then
```
Comment (m-reza-rahman) [inline .github/workflows/nightly.yml:359]: No need for this particular type of check. We should check for expected positive and negative cases.

**Review by m-reza-rahman — CHANGES_REQUESTED** [review]
I think we can trim down some of the somewhat redundant validation cases by taking into account what we are actually validating. Can you take another look from that perspective?

**Comment by Tasnim1147** [comment]
Kept `Run Archetype for EE {8/9/9.1} Core Profile, SE 8` to ensure **Core Profile only available for Jakarta EE 10 and later.**
Kept `Run Archetype for EE 10, SE 8` to ensure **Jakarta EE 10 requires Java SE above 8.** and **Java SE 8 requires Jakarta EE below 10.**
Kept `Run Archetype for EE {9/9.1}, SE 8, WildFly` to ensure **WildFly requires Jakarta EE 8 or 10,**

**Comment by m-reza-rahman** [comment]
OK. Will take a look when I am able.

**Comment by m-reza-rahman** [comment]
Can you resolve the merge conflicts please?

**Comment by Tasnim1147** [comment]
I have resolved the conflicts.

**Review by m-reza-rahman — CHANGES_REQUESTED** [review]
Can you please investigate looking for the actual error message from the Archetype? That will make these tests a bit more robust and also easier to understand/maintain. Not checking that makes it possible that we may actually check for the wrong failure.

File: .github/workflows/nightly.yml [hunk 7]
```diff
@@ -36,6 +36,21 @@ jobs:
...
 
+
+      - name: Run Archetype for EE 8 Core Profile, SE 8
+        continue-on-error: true
+        run: |
+          mvn archetype:generate -DinteractiveMode=false -DaskForDefaultPropertyValues=false -DarchetypeGroupId=org.eclipse.starter -DarchetypeArtifactId=jakarta-starter -DarchetypeVersion=2.3.0-SNAPSHOT -DjakartaVersion=8 -Dprofile=core -DjavaVersion=8 -DoutputDirectory=app -Dgoals="clean package" | tee mvn_output.txt
+
+      - name: Check Maven Build Status
+        run: |
+          ERROR_MESSAGE="Failed to execute goal org.apache.maven.plugins:maven-archetype-plugin:3.2.1:generate (default-cli) on project jakarta-starter-modules: Failed, the Core Profile is only supported for Jakarta EE 10"
```
Comment (m-reza-rahman) [inline .github/workflows/nightly.yml:47]: Can you please shorten the error messages to something like "Core Profile is only supported for Jakarta EE 10"? Otherwise this will be a bit too brittle of a test. The same for the other ones.

**Review by m-reza-rahman — CHANGES_REQUESTED** [review]
Please address feedback.

File: .github/workflows/nightly.yml (hunk 7 above)
Comment (Tasnim1147) [reply .github/workflows/nightly.yml]: Could you please clarify why it's necessary to check for the failure status when we are already checking for the error message, which should indicate that the build failed?

Comment (m-reza-rahman) [reply .github/workflows/nightly.yml]: It's subjective, but the main issue is readability/maintainability. If we explicitly check for failure status, it's semantically clearer to a maintainer what we are testing for. There is also an edge case (less of a concern) that we generate an error message in the Archetype but don't actually fail the build (which in case of Docker is actually valid and what we should check for in the test). Does that help?

Comment (Tasnim1147) [reply .github/workflows/nightly.yml]: Yes. Thank you for the clarification.

File: .github/workflows/nightly.yml [hunk 8]
```diff
@@ -36,6 +36,29 @@ jobs:
...
           rm -rf app
 
+
+      - name: Run Archetype for EE 8 Core Profile, SE 8
+        id: ee_8_core_se_8_none
+        continue-on-error: true
+        run: |
+          mvn archetype:generate -DinteractiveMode=false -DaskForDefaultPropertyValues=false -DarchetypeGroupId=org.eclipse.starter -DarchetypeArtifactId=jakarta-starter -DarchetypeVersion=2.3.0-SNAPSHOT -DjakartaVersion=8 -Dprofile=core -DjavaVersion=8 -DoutputDirectory=app -Dgoals="clean package" > mvn_output.txt 2>&1
+
+      - name: Check Maven Build Status
```
Comment (m-reza-rahman) [inline .github/workflows/nightly.yml:46]: Please see if both of the following checks can be combined. From a business logic and readability perspective, they are inseparable. We are checking that the build fails for a specific reason.

**Review by m-reza-rahman — CHANGES_REQUESTED** [review]
Please address review comments.

**Comment by Tasnim1147** [comment]
I have combined the Check Maven Build Status and Check Error Message steps into a single step using a series of conditionals. However, I am unsure if it is possible to make these checks more concise.

**Comment by m-reza-rahman** [comment]
Please condense both checks in a single if statement separated by a logical AND. This is possible in bash.

**Comment by m-reza-rahman** [comment]
Please double check if “2>&1” is really necessary. I think all the Maven output is in stout. I don’t think combining stderr is needed.

**Comment by m-reza-rahman** [comment]
One more thin to investigate is whether two actions and "continue-on-error" is really needed. It may be possible to do the entire test in one shell session by checking both the Maven exit status and output.

**Comment by Tasnim1147** [comment]
> Please condense both checks in a single if statement separated by a logical AND. This is possible in bash.

If we combine the checks using a logical AND, wouldn't it make it more difficult to identify the exact cause of the failure? For instance, whether it was due to the wrong error message or an incorrect build status.

**Comment by Tasnim1147** [comment]
> Please double check if “2>&1” is really necessary. I think all the Maven output is in stout. I don’t think combining stderr is needed.

Initially, I assumed that the outputs were in both stdout and stderr. However, after rechecking, I found that the output is indeed only in stdout. Thank you for pointing that out.

**Comment by m-reza-rahman** [comment]
> > Please condense both checks in a single if statement separated by a logical AND. This is possible in bash.
> 
> If we combine the checks using a logical AND, wouldn't it make it more difficult to identify the exact cause of the failure? For instance, whether it was due to the wrong error message or an incorrect build status.

A logical AND will only succeed if both conditions are true - the build failed and we found the error message we are looking for. If things failed for some other reason, the build should fail.

**Comment by Tasnim1147** [comment]
Could you please review the following sample to ensure it meets your expectations?

```
      - name: Run Archetype for EE 8 Core Profile, SE 8
        run: |
          mvn archetype:generate -DinteractiveMode=false -DaskForDefaultPropertyValues=false -DarchetypeGroupId=org.eclipse.starter -DarchetypeArtifactId=jakarta-starter -DarchetypeVersion=2.3.0-SNAPSHOT -DjakartaVersion=8 -Dprofile=core -DjavaVersion=8 -DoutputDirectory=app -Dgoals="clean package" | tee mvn_output.txt
          
          MAVEN_EXIT_CODE=${PIPESTATUS[0]}
          ERROR_MESSAGE="Core Profile is only supported for Jakarta EE 10"
          if [ $MAVEN_EXIT_CODE -ne 0 ] && grep -q "$ERROR_MESSAGE" mvn_output.txt; then
              echo "Maven build fail and expected error message was found."
          else
              echo "Maven build did not fail, or the expected error message was not found."
              exit 1
          fi
          
          rm -f mvn_output.txt
 ```

**Comment by m-reza-rahman** [comment]
> Could you please review the following sample to ensure it meets your expectations?
> 
> ```
>       - name: Run Archetype for EE 8 Core Profile, SE 8
>         run: |
>           mvn archetype:generate -DinteractiveMode=false -DaskForDefaultPropertyValues=false -DarchetypeGroupId=org.eclipse.starter -DarchetypeArtifactId=jakarta-starter -DarchetypeVersion=2.3.0-SNAPSHOT -DjakartaVersion=8 -Dprofile=core -DjavaVersion=8 -DoutputDirectory=app -Dgoals="clean package" | tee mvn_output.txt
>           
>           MAVEN_EXIT_CODE=${PIPESTATUS[0]}
>           ERROR_MESSAGE="Core Profile is only supported for Jakarta EE 10"
>           if [ $MAVEN_EXIT_CODE -ne 0 ] && grep -q "$ERROR_MESSAGE" mvn_output.txt; then
>               echo "Maven build fail and expected error message was found."
>           else
>               echo "Maven build did not fail, or the expected error message was not found."
>               exit 1
>           fi
>           
>           rm -f mvn_output.txt
> ```

Very close, but not quite. Firstly, the echo statements are unnecessary if the logic is clear enough. I would remove those. You need to indicate failure if anything else other than exactly what we expect happens. So basically, you need to wrap the conditions you are checking with a logical outer negation.

File: .github/workflows/nightly.yml [hunk 9]
```diff
@@ -920,11 +999,6 @@ jobs:
           docker rmi test-image
           rm -rf app/tomee
 
-      - name: Run Archetype for EE 10 Web Profile, SE 17, WildFly
```
Comment (m-reza-rahman) [inline .github/workflows/nightly.yml:923]: Trying to understand why this was removed. Is it redundant?

File: .github/workflows/nightly.yml
Comment (m-reza-rahman) [inline .github/workflows/nightly.yml:1]: This looks much better, but there seems to be a number of cases missing. Please cross-check here for "none" and WildFly: https://github.com/eclipse-ee4j/starter/blob/master/archetype/src/main/resources/META-INF/archetype-post-generate.groovy. I can see at least these cases:

* Trying to use Docker when no runtime is selected (please check for build success and no Dockerfile - optionally the warning message too).
* Invalid Jakarta EE version (you can use 11).
* Invalid Java SE version (you can use 7).
* Invalid Profile (you can use "microprofile").
* Invalid Docker option (you can use "Y").
* Invalid runtime (you can use "weblogic").

**Review by m-reza-rahman — CHANGES_REQUESTED** [review]
Please resolve/address review comments.

File: .github/workflows/nightly.yml (hunk 9 above)
Comment (Tasnim1147) [reply .github/workflows/nightly.yml:923]: This case was not removed; I simply relocated it to the correct location.

File: .github/workflows/nightly.yml
Comment (Tasnim1147) [reply .github/workflows/nightly.yml:1]: Actually, I was generating negative cases based on the possible combinations of the parameters in [UI](https://start.jakarta.ee/). I didn't know that I had to find the cases based on the [groovy script](https://github.com/eclipse-ee4j/starter/blob/master/archetype/src/main/resources/META-INF/archetype-post-generate.groovy). Thank you for this information.

File: .github/workflows/nightly.yml (hunk 9 above)
Comment (m-reza-rahman) [reply .github/workflows/nightly.yml:923]: I am thinking the diffs may be messed up now. I don’t see a corresponding addition. You may want to start a fresh PR?

Comment (Tasnim1147) [reply .github/workflows/nightly.yml:923]: Sure. #318 will be a continuation of this PR (#310)
